apm-agent-core/src/main/java/co/elastic/apm/agent/logging/ApmServerLogAppender.java

```diff
@@ -0,0 +1,146 @@
+
+    private final List<LogEvent> buffer;
+
+    private ApmServerLogAppender(String name, Layout<?> layout) {
+        super(name, null, layout, true, null);
+        this.buffer = new ArrayList<>();
+    }
+
+    public static ApmServerLogAppender getInstance() {
+        return Objects.requireNonNull(INSTANCE);
+    }
+
+    @SuppressWarnings("unused")
+    @PluginFactory
+    public static ApmServerLogAppender createAppender(@PluginAttribute("name") String name,
+                                                      @PluginElement("Layout") Layout<?> layout) {
+
+        if(!(layout instanceof EcsLayout)){
+            throw new IllegalArgumentException("invalid layout "+ layout);
+        }
+
+        if (INSTANCE == null) {
+            INSTANCE = new ApmServerLogAppender(name, layout);
+        }
+
+        return INSTANCE;
+    }
+
+    @Override
+    public void append(LogEvent event) {
+        if (!isAgentInitialized()) {
+            synchronized (buffer) {
+                if (buffer.size() < MAX_BUFFER_SIZE) {
+                    buffer.add(event.toImmutable());
+                }
+            }
+            return;
+        }
+
+        if (!config.getSendLogs()) {
+            return;
+        }
+        sendLogEvent(event);
+    }
+
+    public LifecycleListener getInitListener() {
+        return new AbstractLifecycleListener() {
+            @Override
+            public void init(ElasticApmTracer tracer) throws Exception {
+                initStreaming(tracer.getConfig(LoggingConfiguration.class), tracer.getReporter());
+            }
+        };
+    }
+
+    private void initStreaming(LoggingConfiguration config, Reporter reporter) {
+        if (isAgentInitialized()) {
+            throw new IllegalStateException("streaming already initialized");
+        }
+
+        this.config = config;
+        this.reporter = reporter;
+
+        synchronized (buffer) {
+            for (LogEvent logEvent : buffer) {
+                sendLogEvent(logEvent);
+            }
+            buffer.clear();
+        }
+    }
+
+
+    private void sendLogEvent(LogEvent event) {
+        // When trying to debug wire protocol, adding the whole HTTP request body nested in a log message is not possible
+        // otherwise it makes the agent recursively nest payload data until the request size limit is reached.
+        //
+        // We have to filter this exception here to still provide the ability to log to filesystem if needed.
+        if (event.getLevel().intLevel() >= Level.TRACE.intLevel() && event.getLoggerName().equals(DslJsonSerializer.class.getName())) {
+            MutableLogEvent newEvent = new MutableLogEvent();
+            newEvent.initFrom(event);
+            newEvent.setMessage(new SimpleMessage("wire protocol logging only available when logging to filesystem"));
+            event = newEvent;
+        }
+
+        Objects.requireNonNull(reporter).reportAgentLog(getLayout().toByteArray(event));
```

Comment:
It's just a way to avoid a warning in the IDE, while it will never be `null` in practice, we could also use an `assert reporter != null` if you think that would make it clearer here.

Reply:
Ah, I see. According to [this SO response](https://stackoverflow.com/a/42127085) we could also just suppress the IDE warning on this line:

```
//noinspection ConstantConditions
reporter.reportAgentLog(getLayout().toByteArray(event));
```

But it doesn't really change much here, so you can just leave it as is if you want to.